lib/omnibus/health_check.rb

```diff
@@ -399,10 +477,40 @@ def whitelist_files
     # @yield [String]
     #   each line
     #
-    def read_shared_libs(command)
-      cmd = shellout(command)
-      cmd.stdout.each_line do |line|
-        yield line
+    def read_shared_libs(find_command, ldd_command, &output_proc)
+      #
+      # construct the list of files to check
+      #
+
+      find_output = shellout!(find_command).stdout.lines
+
+      find_output.reject! { |file| IGNORED_ENDINGS.any? { |ending| file.end_with?("#{ending}\n") } }
+
+      find_output.reject! { |file| IGNORED_SUBSTRINGS.any? { |substr| file.include?(substr) } }
+
+      if find_output.empty?
+        # probably the find_command is busted, it should never be empty or why are you using omnibus?
+        raise "Internal Error: Health Check found no lines"
+      end
+
+      if find_output.any? { |file| file !~ Regexp.new(project.install_dir) }
+        # every file in the find output should be within the install_dir
+        raise "Internal Error: Health Check lines not matching the install_dir"
+      end
+
+      #
+      # feed the list of files to the "ldd" command
+      #
+
+      # this command will typically fail if the last file isn't a valid lib/binary which happens often
```

Comment:
I believe I may have spotted a regression; or I'm just hitting a non-deterministic issue more frequently with these code changes

The scenario is - I've just upgraded to a newer version of Omnibus, and it looks like the `ldd` command here can break halfway through with this current pattern. It took a while to track down things as the status code here is getting ignored, so the error was being swallowed.

Example, reading 3 files and the 2nd ELF file causes an error:

```
[9] pry(#<Omnibus::HealthCheck>)> puts ldd_output = shellout(ldd_command, input: "/opt/project/file1\n/opt/project/file2\n/opt/project/file3").stdout
/opt/project/file1:
	not a dynamic executable
/opt/project/file2:
=> nil
```

We get a partial result in the current implementation, but if you extract the status code you can see it has failed an exitstatus of 135:

```
[11] pry(#<Omnibus::HealthCheck>)> puts ldd_output = shellout(ldd_command, input: "/opt/project/file1\n/opt/project/file2\n/opt/project/file3").result
NoMethodError: undefined method `result' for <Mixlib::ShellOut#1230: command: 'xargs ldd' process_status: #<Process::Status: pid 46708 exit 123> stdout: '/opt/project/file1:
	not a dynamic executable
/opt/project/file2:' stderr: 'ldd: exited with unknown exit code (135)' child_pid: 46708 environment: {} timeout: 7200 user:  group:  working_dir:  >:Mixlib::ShellOut
from (pry):11:in `block in read_shared_libs'
```

I actually thought there was a bug in `shellout(...).stdout` since it looked like partial reads of stdout.

For my current setup, this new code path skips multiple healthchecks as a result of exiting earlier than expected

Reply:
Initial pull request: https://github.com/chef/omnibus/pull/1137